VPD parser crashes in some cases (case 3)

<!--
Github issues are used for bug reports. For support questions, please use [our forum](https://discuss.linuxcontainers.org).

Please fill the template below as it will greatly help us track down your issue and reproduce it on our side.  
Feel free to remove anything which doesn't apply to you and add more information where it makes sense.
-->

# Required information

 * Distribution: Ubuntu
 * Distribution version: 20.04 (not specific to distro)
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.4.0-121-generic (not specific to kernel)
   * LXC version: 5.0.0
   * LXD version: b2a75dfedc8cc26e0c58ab20164af7c3aa84b14b (MAAS 3.2.0-11998-g.3e666807a)

# Issue description

Another VPD parser crash, as in #10627 and #10662.

Relevant vpd data in `/sys/bus/pci/devices/0000:02:00.0/vpd`:
[bad_vpd_02_00.0.zip](https://github.com/lxc/lxd/files/9212257/bad_vpd_02_00.0.zip)



```
$ sudo strace -f /snap/maas/22919/usr/share/maas/machine-resources/amd64 2>&1 | grep /vpd
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:00.0/vpd", 0xc000bf8108, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:01.0/vpd", 0xc000bf92e8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:02.0/vpd",  <unfinished ...>
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:03.0/vpd", 0xc000c016f8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:04.0/vpd", 0xc000c08928, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:05.0/vpd", 0xc000c09b08, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:06.0/vpd", 0xc000c14d38, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:07.0/vpd", 0xc000c15f18, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:08.0/vpd", 0xc000c1f148, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:09.0/vpd", 0xc000c20378, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0a.0/vpd", 0xc000c21558, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0d.0/vpd", 0xc000c21ca8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0d.1/vpd", 0xc000c2c448, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0d.2/vpd", 0xc000c2cb98, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0d.3/vpd", 0xc000c2d2e8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0d.4/vpd", 0xc000c2da38, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0d.5/vpd", 0xc000c2e108, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0d.6/vpd", 0xc000c2e858, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0e.0/vpd",  <unfinished ...>
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0e.1/vpd", 0xc000c2f6f8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0e.2/vpd", 0xc000c2fe48, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0e.3/vpd", 0xc000cac5e8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:0e.4/vpd", 0xc000cacd38, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:14.0/vpd", 0xc000cadf18, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:14.1/vpd", 0xc000cba6b8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:14.2/vpd", 0xc000cbae08, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1c.0/vpd", 0xc000cc2038, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1c.2/vpd", 0xc000cc3218, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1d.0/vpd", 0xc000ccc448, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1d.1/vpd", 0xc000ccd628, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1d.2/vpd", 0xc000cd6858, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1d.3/vpd", 0xc000cd7a38, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1d.7/vpd", 0xc000cdec68, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1e.0/vpd", 0xc000cdf3b8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1f.0/vpd", 0xc0003e0378, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:01:03.0/vpd", 0xc0003e1628, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:01:04.0/vpd", 0xc0003e1d78, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:01:04.2/vpd", 0xc000c4f218, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:01:04.4/vpd",  <unfinished ...>
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:01:04.6/vpd", 0xc00010c518, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 2366039] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:02:00.0/vpd", {st_mode=S_IFREG|0600, st_size=0, ...}, AT_SYMLINK_NOFOLLOW) = 0
[pid 2366039] openat(AT_FDCWD, "/sys/bus/pci/devices/0000:02:00.0/vpd", O_RDONLY|O_CLOEXEC) = 3
```

```
$ sudo /snap/maas/22919/usr/share/maas/machine-resources/amd64 
panic: runtime error: slice bounds out of range [1:0]

goroutine 1 [running]:
github.com/lxc/lxd/lxd/resources.vpdReadInt(...)
        /build/maas/parts/host-info/build/vendor/github.com/lxc/lxd/lxd/resources/pci_vpd.go:44
github.com/lxc/lxd/lxd/resources.vpdReadEntries({0xc0009ac035, 0x4b, 0x1cb}, 0x4a)
        /build/maas/parts/host-info/build/vendor/github.com/lxc/lxd/lxd/resources/pci_vpd.go:68 +0x337
github.com/lxc/lxd/lxd/resources.parsePCIVPD({0xc0009ac000, 0x80, 0x200})
        /build/maas/parts/host-info/build/vendor/github.com/lxc/lxd/lxd/resources/pci_vpd.go:115 +0x136
github.com/lxc/lxd/lxd/resources.GetPCI()
        /build/maas/parts/host-info/build/vendor/github.com/lxc/lxd/lxd/resources/pci.go:146 +0x9dc
github.com/lxc/lxd/lxd/resources.GetResources()
        /build/maas/parts/host-info/build/vendor/github.com/lxc/lxd/lxd/resources/resources.go:48 +0x267
host-info/pkg/info.GetResources(...)
        /build/maas/parts/host-info/build/pkg/info/info.go:157
host-info/pkg/info.GetInfo()
        /build/maas/parts/host-info/build/pkg/info/info.go:182 +0x34
main.main()
        /build/maas/parts/host-info/build/cmd/machine-resources/main.go:23 +0x1d
```

--- a/lxd/resources/pci_vpd.go
+++ b/lxd/resources/pci_vpd.go
@@ -36,6 +36,10 @@ func vpdKnownKey(name string) bool {
 }
 
 func vpdReadInt(buf []byte, length int) ([]byte, int) {
+	if length > len(buf) {
+		return []byte{}, 0
+	}
+
 	value := 0
 	for i, n := range buf[:length] {
 		value += int(n) << (i * 8)

